### ISSUE TITLE
Node is exiting before http request finishes

### ISSUE THREAD

**Issue by georgehdd** [TooTallNate/node-agent-base#7]
My organization is using typings@0.6.8 in gulp in order to install typescript definitions for 3rd parties.
As of the latest change to agent-base (2.1.0), gulp (node) is exiting before typings job finishes downloading all the .d.ts.
After some debugging I found out that what is causing the issue is the following commit:
https://github.com/TooTallNate/node-agent-base/commit/7858f43a6b71861ec11d282bc17a83cf9444c9f8

This module is consumed by the following chain: typings@0.6.8 => popsicle@3.2.2 => popsicle-proxy-agent@1.0.0 => http-proxy-agent@1.0.0 => agent-base@2.1.0

Attached is an example which reproduces the issue we're encountering. Simply download, extract, run **npm install** and then **npm start**: [issue-reproduce.zip](https://github.com/TooTallNate/node-agent-base/files/1034214/issue-reproduce.zip)

**Comment by dweber019** [comment]
We have the same issue.
We are using Jenkins and this defect is causing the typings to not install because other scripts are executed in the chain.

**Comment by clcpolevaulter** [comment]
This is also failing our request through the agent when used with KMS. This was working perfectly with version 2.0.1 of agent-base. Please consider releasing a patch update of https://github.com/TooTallNate/node-proxy-agent that locks the agent-base until further investigation can be done. 

``` js

import AWS from 'aws-sdk';
import proxy from 'proxy-agent';
import config from 'config';

function configureProxy() {
    if (config.proxy) {
        AWS.config.update({
            httpOptions: {
                agent: proxy(config.proxy)
            }
        });
    }
}

export default {
    decrypt(encryptedBlob, callback) {
        let kms;

        configureProxy();

        kms = new AWS.KMS({region: 'us-east-1'});

        kms.decrypt(
            {
                CiphertextBlob: new Buffer(encryptedBlob, 'base64')
            },
            (err, data) => {
                if (err) {
                    callback(err);
                } else {
                    callback(undefined, data.Plaintext);
                }
            }
        );
    }
};
```

**Comment by pi0** [comment]
Apologizes for any inconveniences caused by this commit. I confirm too. It seems making problems with processes having only one callback waiting for socket. I will create a hot-fix PR keeping users safe from this change. This behavior then can be enabled using environment variables.

**Comment by TooTallNate** [comment]
> This behavior then can be enabled using environment variables.

There shouldn't be any opt-in behavior like that with this module. It should behave transparently just like the default `http.Agent` in Node core works.

**Comment by pi0** [comment]
@TooTallNate I was trying to use this opt-in option as a temporary workaround until we found a better way. If you prefer we may just revert everything back to original state :(

**Comment by clcpolevaulter** [comment]
@TooTallNate thoughts on #8?

**Comment by TooTallNate** [comment]
Reverted the patch in https://github.com/TooTallNate/node-agent-base/commit/bda8cf12f07462329709759bcc96cf13d37b2813 (sorry @pi0!).

Published in v2.1.1.

**Comment by clcpolevaulter** [comment]
thanks @TooTallNate
